tests/test_dulwich.py

```diff
@@ -26,3 +30,21 @@ def test_dulwich_github_compat(mocker: MockerFixture, algorithm: bytes):
     strings = iter((b"ssh-rsa", key_data))
     packet.get_string = lambda: next(strings)
     _process_public_key_ok_gh(auth, None, None, packet)
+
+
+def test_dulwich_get_ref(tmp_dir: TmpDir, scm: Git):
+    backend = DulwichBackend(tmp_dir)
+    tmp_dir.gen("foo", "foo")
+    scm.add_commit("foo", message="foo")
+
+    scm.set_ref("refs/exp/base", "refs/heads/master")
```

Comment:
```suggestion
    scm.set_ref("refs/exp/base", "refs/heads/master", symbolic=True)
```

Reply:
It's returning the newline because the ref created in line 40 is not defined as a symbolic reference.

A symbolic reference file contains a special header (`ref: `) and then the target ref string like:
```
ref: refs/target/ref
```
While a non-symbolic reference file only contains a single 40-byte git SHA like:
```
d670460b4b4aece5915caf5c68d12f560a9fe3e4
```

If you call `set_ref` without the symbolic flag, whatever you pass as the ref target gets written directly to the file (in this case `refs/heads/master` with no symbolic ref header).

When we call `get_ref`, dulwich looks for the symbolic ref header `ref: `, and since that header wasn't created, dulwich treats the it as a regular reference and just returns the first 40 bytes in that file (which would normally be the git SHA and no newline).

Reply:
Got it. I just copied from https://github.com/iterative/dvc/blob/master/tests/func/experiments/test_experiments.py#L632 to add a test case replicating the failures on the DVC PR:

https://github.com/iterative/dvc/runs/4302184106?check_suite_focus=true